src/main/java/com/es/phoneshop/dao/ArrayListProductDao.java

```diff
@@ -129,6 +137,37 @@ private void updateProducts(Product product) {
                 );
     }
 
+
+    public Deque<Long> getRecentProductIds(HttpServletRequest request) {
+        HttpSession session = request.getSession();
+        synchronized (session) {
+            Deque<Long> recentProductIds = (Deque<Long>) session.getAttribute(RECENT_SESSION_ATTR);
+            if (recentProductIds == null) {
+                session.setAttribute(RECENT_SESSION_ATTR, recentProductIds = new ArrayDeque<>());
+            }
+            return recentProductIds;
+        }
+    }
+
+
+    @Override
+    public List<Product> getRecentProducts(int count, HttpServletRequest request) {
+        Deque<Long> recentProductIds = getRecentProductIds(request);
+        return recentProductIds.stream()
+                .distinct()
+                .limit(count)
+                .map(this::getProduct)
+                .collect(Collectors.toList());
+    }
+
+
+    @Override
+    public void addToRecent(Long id, Deque<Long> recentProductIds) {
+        writeLock.lock();
+        recentProductIds.addFirst(id);
+        writeLock.unlock();
```

Comment:
Do you really need lock here?

Reply:
It might be good idea to limit amount of products that stored in recentProductsIds